json_serializable/lib/src/utils.dart

```diff
@@ -38,5 +51,182 @@ String friendlyNameForElement(Element element) {
   return names.join(' ');
 }
 
-final toJsonMapVarName = 'val';
-final toJsonMapHelperName = 'writeNotNull';
+/// Returns a list of all instance, [FieldElement] items for [element] and
+/// super classes.
+List<FieldElement> listFields(ClassElement element) {
+  // Get all of the fields that need to be assigned
+  // TODO: support overriding the field set with an annotation option
+  var fieldsList = element.fields.where((e) => !e.isStatic).toList();
+
+  var manager = new InheritanceManager(element.library);
+
+  for (var v in manager.getMembersInheritedFromClasses(element).values) {
+    assert(v is! FieldElement);
+    if (_dartCoreObjectChecker.isExactly(v.enclosingElement)) {
+      continue;
+    }
+
+    if (v is PropertyAccessorElement && v.variable is FieldElement) {
+      fieldsList.add(v.variable as FieldElement);
+    }
+  }
+
+  var undefinedFields = fieldsList.where((fe) => fe.type.isUndefined).toList();
+  if (undefinedFields.isNotEmpty) {
+    var description =
+        undefinedFields.map((fe) => '`${fe.displayName}`').join(', ');
+
+    throw new InvalidGenerationSourceError(
+        'At least one field has an invalid type: $description.',
+        todo: 'Check names and imports.');
+  }
+
+  // Sort these in the order in which they appear in the class
+  // Sadly, `classElement.fields` puts properties after fields
+  fieldsList.sort(_sortByLocation);
+
+  return fieldsList;
+}
+
+int _sortByLocation(FieldElement a, FieldElement b) {
+  var checkerA = new TypeChecker.fromStatic(a.enclosingElement.type);
+
+  if (!checkerA.isExactly(b.enclosingElement)) {
+    // in this case, you want to prioritize the enclosingElement that is more
+    // "super".
+
+    if (checkerA.isSuperOf(b.enclosingElement)) {
+      return -1;
+    }
+
+    var checkerB = new TypeChecker.fromStatic(b.enclosingElement.type);
+
+    if (checkerB.isSuperOf(a.enclosingElement)) {
+      return 1;
+    }
+  }
+
+  /// Returns the offset of given field/property in its source file – with a
+  /// preference for the getter if it's defined.
+  int _offsetFor(FieldElement e) {
+    if (e.getter != null && e.getter.nameOffset != e.nameOffset) {
+      assert(e.nameOffset == -1);
+      return e.getter.nameOffset;
+    }
+    return e.nameOffset;
+  }
+
+  return _offsetFor(a).compareTo(_offsetFor(b));
+}
+
+final _dartCoreObjectChecker = const TypeChecker.fromRuntime(Object);
+
+/// Returns the set of fields that are written to via factory.
```

Comment:
blank line after the first sentence

I'm assuming there are side effects to this method (like writing to `buffer`) - maybe we should focus on those in the first description rather than the return value?

Reply:
Did a big clean-up here